src/fragment-generation-utils.js

```diff
@@ -953,6 +964,81 @@ const FragmentFactory = class {
+   */
+  constructor(searchRange, isForwardTraversal) {
+    this.searchRange = searchRange;
+    this.insertFunc = isForwardTraversal ? Array.prototype.push : Array.prototype.unshift;
+    this.textFound = false;
+    this.textNodes = [];
+    this.textInBlock = null;
+  }
+  /**
+   * Adds the next node in the search space range traversal to the accumulator.
+   * The accumulator then will keep track of the text nodes in the range until a block boundary is found.
+   * Once a block boundary is found and the content of the text nodes in the boundary is non emtpy, the property
+   * textInBlock will be set with the content of the text nodes, trimmed of leading and trailing whitespaces.
+   * @param {Node} node - next node in the traversal of the searchRange  
+   */
+  appendNode(node) {
+    // if we already calculated the text in the block boundary just ignore any calls to append nodes
+    if (this.textInBlock !== null) {
+      return;
+    }
+    // we found a block boundary, check if there's text inside and set it to textInBlock or keep going to the next block boundary
+    if (isBlock(node)) {
+      if (this.textFound) {
+        // concatenate all the text nodes in the block boundary and trim any trailing and leading whitespaces
+        this.textInBlock = this.textNodes.map(textNode => textNode.textContent).join("").trim();
+      } else {
+        // discard the text nodes visited so far since they are empty and we'll continue searching in the next block boundary
+        this.textNodes = [];
+      }
+      return;
+    }
+
+    // ignore non text nodes
+    if (!isText(node)) return;
+
+    // get the part of node inside the search range. this is to avoid accumulating text that's not inside the range
+    const nodeToInsert = this.getNodeInterceptionWithRange(node);
+
+    // keep track of any text found in the block boundary.
+    this.textFound = this.textFound || nodeToInsert.textContent.trim() !== "";
+
+    this.insertFunc.call(this.textNodes, nodeToInsert);
+  }
+
+  /**
+   * Calculates the interception of a node with searchRange and returns a Text Node with the interception
+   * @param {Node} node - the node to intercept with searchRange
+   * @returns {Node} - node if node is fully within searchRange or a Text Node with the substring of the content of node inside the search range
+   */
+  getNodeInterceptionWithRange(node) {
+    let startOffset = null;
+    let endOffset = null;
+
+    if (node === this.searchRange.startContainer && this.searchRange.startOffset !== 0) {
+      startOffset = this.searchRange.startOffset;
+    }
+
+    if (node === this.searchRange.endContainer && this.searchRange.endOffset !== node.textContent.length) {
+      endOffset = this.searchRange.endOffset;
+    }
+    if (startOffset !== null || endOffset !== null) {
+      return document.createTextNode(node.textContent.substring(startOffset ?? 0, endOffset ?? node.textContent.length));
```

Comment:
I still don't love creating a text node that's not really a node (i.e., not part of the document tree). Did you try the idea of just returning `{textContent: node.textContent.substring(startOffset ?? 0, endOffset ?? node.textContent.length)}`?